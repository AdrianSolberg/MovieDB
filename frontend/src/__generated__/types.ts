export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
export type MakeEmpty<T extends { [key: string]: unknown }, K extends keyof T> = { [_ in K]?: never };
export type Incremental<T> = T | { [P in keyof T]?: P extends ' $fragmentName' | '__typename' ? T[P] : never };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: { input: string; output: string; }
  String: { input: string; output: string; }
  Boolean: { input: boolean; output: boolean; }
  Int: { input: number; output: number; }
  Float: { input: number; output: number; }
  Date: { input: any; output: any; }
};

export type Movie = {
  _id: Scalars['Int']['output'];
  title: Scalars['String']['output'];
  vote_average: Scalars['Float']['output'];
  vote_count: Scalars['Int']['output'];
  status: Scalars['String']['output'];
  release_date: Scalars['Date']['output'];
  revenue: Scalars['Float']['output'];
  runtime: Scalars['Int']['output'];
  backdrop_path?: Maybe<Scalars['String']['output']>;
  budget: Scalars['Int']['output'];
  homepage?: Maybe<Scalars['String']['output']>;
  imdb_id?: Maybe<Scalars['String']['output']>;
  original_language: Scalars['String']['output'];
  original_title: Scalars['String']['output'];
  overview: Scalars['String']['output'];
  popularity: Scalars['Float']['output'];
  poster_path?: Maybe<Scalars['String']['output']>;
  tagline?: Maybe<Scalars['String']['output']>;
  genres: Array<Scalars['String']['output']>;
  production_companies: Array<Scalars['String']['output']>;
  production_countries: Array<Scalars['String']['output']>;
  spoken_languages: Array<Scalars['String']['output']>;
  keywords: Array<Scalars['String']['output']>;
};

export type Filters = {
  Genre: Array<Scalars['String']['output']>;
  Rating: Array<Scalars['String']['output']>;
  Decade: Array<Scalars['String']['output']>;
  Status: Array<Scalars['String']['output']>;
  Runtime: Array<Scalars['String']['output']>;
};

export type FiltersInput = {
  Genre: Array<Scalars['String']['input']>;
  Rating: Array<Scalars['String']['input']>;
  Decade: Array<Scalars['String']['input']>;
  Status: Array<Scalars['String']['input']>;
  Runtime: Array<Scalars['String']['input']>;
};

export enum SortingType {
  NewestFirst = 'NEWEST_FIRST',
  OldestFirst = 'OLDEST_FIRST',
  BestRated = 'BEST_RATED',
  WorstRated = 'WORST_RATED',
  LongestRuntime = 'LONGEST_RUNTIME',
  ShortestRuntime = 'SHORTEST_RUNTIME'
}

export type Query = {
  movie?: Maybe<Movie>;
  movies: Array<Movie>;
  filters?: Maybe<Filters>;
};


export type QueryMovieArgs = {
  id: Scalars['Int']['input'];
};


export type QueryMoviesArgs = {
  skip?: InputMaybe<Scalars['Int']['input']>;
  limit?: InputMaybe<Scalars['Int']['input']>;
  filters?: InputMaybe<FiltersInput>;
  sortOption?: InputMaybe<SortingType>;
  search?: InputMaybe<Scalars['String']['input']>;
};

export type GetMovieQueryVariables = Exact<{
  movieId: Scalars['Int']['input'];
}>;


export type GetMovieQuery = { __typename?: 'Query', movie?: { __typename?: 'Movie', _id: number, title: string, vote_average: number, vote_count: number, status: string, release_date: any, revenue: number, runtime: number, backdrop_path?: string | null, budget: number, homepage?: string | null, imdb_id?: string | null, original_language: string, original_title: string, overview: string, popularity: number, poster_path?: string | null, tagline?: string | null, genres: Array<string>, production_companies: Array<string>, production_countries: Array<string>, spoken_languages: Array<string>, keywords: Array<string> } | null };

export type GetMoviesQueryVariables = Exact<{
  skip?: InputMaybe<Scalars['Int']['input']>;
  limit?: InputMaybe<Scalars['Int']['input']>;
  filters?: InputMaybe<FiltersInput>;
  sortOption?: InputMaybe<SortingType>;
  search?: InputMaybe<Scalars['String']['input']>;
}>;


export type GetMoviesQuery = { __typename?: 'Query', movies: Array<{ __typename?: 'Movie', _id: number, title: string, vote_average: number, release_date: any, runtime: number, poster_path?: string | null }> };

export type GetFiltersQueryVariables = Exact<{ [key: string]: never; }>;


export type GetFiltersQuery = { __typename?: 'Query', filters?: { __typename?: 'Filters', Genre: Array<string>, Rating: Array<string>, Decade: Array<string>, Status: Array<string>, Runtime: Array<string> } | null };
