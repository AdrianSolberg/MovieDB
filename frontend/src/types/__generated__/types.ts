export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = {
    [K in keyof T]: T[K];
};
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & {
    [SubKey in K]?: Maybe<T[SubKey]>;
};
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & {
    [SubKey in K]: Maybe<T[SubKey]>;
};
export type MakeEmpty<
    T extends { [key: string]: unknown },
    K extends keyof T,
> = { [_ in K]?: never };
export type Incremental<T> =
    | T
    | {
          [P in keyof T]?: P extends " $fragmentName" | "__typename"
              ? T[P]
              : never;
      };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
    ID: { input: string; output: string };
    String: { input: string; output: string };
    Boolean: { input: boolean; output: boolean };
    Int: { input: number; output: number };
    Float: { input: number; output: number };
    DateTime: { input: Date; output: Date };
};

export type Movie = {
    __typename?: "Movie";
    _id: Scalars["Int"]["output"];
    title: Scalars["String"]["output"];
    vote_average: Scalars["Float"]["output"];
    vote_count: Scalars["Int"]["output"];
    status: Scalars["String"]["output"];
    release_date: Scalars["DateTime"]["output"];
    revenue: Scalars["Float"]["output"];
    runtime: Scalars["Int"]["output"];
    backdrop_path?: Maybe<Scalars["String"]["output"]>;
    budget: Scalars["Int"]["output"];
    homepage?: Maybe<Scalars["String"]["output"]>;
    imdb_id?: Maybe<Scalars["String"]["output"]>;
    original_language: Scalars["String"]["output"];
    original_title: Scalars["String"]["output"];
    overview: Scalars["String"]["output"];
    popularity: Scalars["Float"]["output"];
    poster_path?: Maybe<Scalars["String"]["output"]>;
    tagline?: Maybe<Scalars["String"]["output"]>;
    genres: Array<Scalars["String"]["output"]>;
    production_companies: Array<Scalars["String"]["output"]>;
    production_countries: Array<Scalars["String"]["output"]>;
    spoken_languages: Array<Scalars["String"]["output"]>;
    keywords: Array<Scalars["String"]["output"]>;
    reviews: Array<Review>;
};

export type Review = {
    __typename?: "Review";
    _id: Scalars["ID"]["output"];
    movie: Movie;
    username: Scalars["String"]["output"];
    rating: Scalars["Int"]["output"];
    comment: Scalars["String"]["output"];
    date: Scalars["DateTime"]["output"];
};

export type Filters = {
    __typename?: "Filters";
    Genre: Array<Scalars["String"]["output"]>;
    Rating: Array<Scalars["String"]["output"]>;
    Decade: Array<Scalars["String"]["output"]>;
    Status: Array<Scalars["String"]["output"]>;
    Runtime: Array<Scalars["String"]["output"]>;
};

export type FiltersInput = {
    Genre: Array<Scalars["String"]["input"]>;
    Rating: Array<Scalars["String"]["input"]>;
    Decade: Array<Scalars["String"]["input"]>;
    Status: Array<Scalars["String"]["input"]>;
    Runtime: Array<Scalars["String"]["input"]>;
};

export enum SortingType {
    MOST_POPULAR = "MOST_POPULAR",
    NEWEST_FIRST = "NEWEST_FIRST",
    OLDEST_FIRST = "OLDEST_FIRST",
    BEST_RATED = "BEST_RATED",
    WORST_RATED = "WORST_RATED",
    LONGEST_RUNTIME = "LONGEST_RUNTIME",
    SHORTEST_RUNTIME = "SHORTEST_RUNTIME",
}

export type Query = {
    __typename?: "Query";
    movie?: Maybe<Movie>;
    movies: Array<Movie>;
    filters: Filters;
    latestReviews: Array<Review>;
    userReviews: Array<Review>;
};

export type QuerymovieArgs = {
    id: Scalars["Int"]["input"];
};

export type QuerymoviesArgs = {
    skip?: InputMaybe<Scalars["Int"]["input"]>;
    limit?: InputMaybe<Scalars["Int"]["input"]>;
    filters?: InputMaybe<FiltersInput>;
    sortOption?: InputMaybe<SortingType>;
    search?: InputMaybe<Scalars["String"]["input"]>;
};

export type QuerylatestReviewsArgs = {
    skip?: InputMaybe<Scalars["Int"]["input"]>;
    limit?: InputMaybe<Scalars["Int"]["input"]>;
};

export type QueryuserReviewsArgs = {
    username: Scalars["String"]["input"];
    skip?: InputMaybe<Scalars["Int"]["input"]>;
    limit?: InputMaybe<Scalars["Int"]["input"]>;
};

export type Mutation = {
    __typename?: "Mutation";
    addReview: Movie;
    deleteReview: Movie;
};

export type MutationaddReviewArgs = {
    movie_id: Scalars["Int"]["input"];
    username: Scalars["String"]["input"];
    rating: Scalars["Int"]["input"];
    comment: Scalars["String"]["input"];
};

export type MutationdeleteReviewArgs = {
    _id: Scalars["ID"]["input"];
};

export type GetMoviesQueryVariables = Exact<{
    skip?: InputMaybe<Scalars["Int"]["input"]>;
    limit?: InputMaybe<Scalars["Int"]["input"]>;
    filters?: InputMaybe<FiltersInput>;
    sortOption?: InputMaybe<SortingType>;
    search?: InputMaybe<Scalars["String"]["input"]>;
}>;

export type GetMoviesQuery = {
    __typename?: "Query";
    movies: Array<{
        __typename?: "Movie";
        _id: number;
        title: string;
        vote_average: number;
        release_date: Date;
        runtime: number;
        poster_path?: string | null;
    }>;
};

export type GetMovieQueryVariables = Exact<{
    movieId: Scalars["Int"]["input"];
}>;

export type GetMovieQuery = {
    __typename?: "Query";
    movie?: {
        __typename?: "Movie";
        _id: number;
        title: string;
        vote_average: number;
        vote_count: number;
        status: string;
        release_date: Date;
        revenue: number;
        runtime: number;
        backdrop_path?: string | null;
        budget: number;
        homepage?: string | null;
        imdb_id?: string | null;
        original_language: string;
        original_title: string;
        overview: string;
        popularity: number;
        poster_path?: string | null;
        tagline?: string | null;
        genres: Array<string>;
        production_companies: Array<string>;
        production_countries: Array<string>;
        spoken_languages: Array<string>;
        keywords: Array<string>;
        reviews: Array<{
            __typename?: "Review";
            _id: string;
            username: string;
            rating: number;
            comment: string;
            date: Date;
        }>;
    } | null;
};

export type GetFiltersQueryVariables = Exact<{ [key: string]: never }>;

export type GetFiltersQuery = {
    __typename?: "Query";
    filters: {
        __typename?: "Filters";
        Genre: Array<string>;
        Rating: Array<string>;
        Decade: Array<string>;
        Status: Array<string>;
        Runtime: Array<string>;
    };
};

export type AddReviewMutationVariables = Exact<{
    movieId: Scalars["Int"]["input"];
    username: Scalars["String"]["input"];
    rating: Scalars["Int"]["input"];
    comment: Scalars["String"]["input"];
}>;

export type AddReviewMutation = {
    __typename?: "Mutation";
    addReview: {
        __typename?: "Movie";
        reviews: Array<{
            __typename?: "Review";
            _id: string;
            username: string;
            rating: number;
            comment: string;
            date: Date;
        }>;
    };
};
