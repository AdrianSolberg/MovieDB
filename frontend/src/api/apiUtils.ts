import { Review } from "@/types/__generated__/types";

type ResponseReview = {
    __typename?: "Review";
    _id: string;
    username: string;
    rating: number;
    comment: string;
    date: Date;
};

/**
 *  Make sure that the date-field is correctly initialized as a Date-object
 *  when getting reviews as a response from a query.
 */
export function parseReviewResponse(inputReviews: ResponseReview[]): Review[] {
    return inputReviews.map((review) => ({
        ...review,
        date: new Date(review.date),
    })) as Review[];
}
