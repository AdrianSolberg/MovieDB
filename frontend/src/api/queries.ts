import { gql } from "@/types/__generated__";

export const GET_MOVIES = gql(`
    query GetMovies(
        $skip: Int
        $limit: Int
        $filters: FiltersInput
        $sortOption: SortingType
        $search: String
    ) {
        movies(
            skip: $skip
            limit: $limit
            filters: $filters
            sortOption: $sortOption
            search: $search
        ) {
            _id
            title
            vote_average
            release_date
            runtime
            poster_path
        }
    }
`);

export const GET_MOVIE = gql(`
    query GetMovie($movieId: Int!) {
        movie(id: $movieId) {
            _id
            title
            vote_average
            vote_count
            status
            release_date
            revenue
            runtime
            backdrop_path
            budget
            homepage
            imdb_id
            original_language
            original_title
            overview
            popularity
            poster_path
            tagline
            genres
            production_companies
            production_countries
            spoken_languages
            keywords
            reviews {
                _id
                username
                rating
                comment
                date
            }
        }
    }
`);

export const GET_FILTERS = gql(`
    query GetFilters {
        filters {
            Genre
            Rating
            Decade
            Status
            Runtime
        }
    }
`);

export const GET_LATEST_REVIEWS = gql(`
    query GetLatestReviews(
        $skip: Int,
        $limit: Int
    ) {
        latestReviews(
            skip: $skip,
            limit: $limit
        ) {
            _id
            movie {
                _id
                title
                poster_path
            }
            username
            rating
            comment
            date
        }
    }
`);

export const GET_USER_REVIEWS = gql(`
    query GetUserReviews(
        $username: String!,
        $skip: Int,
        $limit: Int
    ) {
        userReviews(
            username: $username,
            skip: $skip,
            limit: $limit
        ) {
            _id
            movie {
                _id
                title
                poster_path
            }
            rating
            comment
            date
        }
    }
`);

export const ADD_REVIEW = gql(`
    mutation AddReview(
        $movieId: Int!,
        $username: String!,
        $rating: Int!,
        $comment: String!
    ) {
        addReview(
            movie_id: $movieId,
            username: $username,
            rating: $rating,
            comment: $comment
        ) {
            _id
            vote_average
            vote_count
            reviews {
                _id
                username
                rating
                comment
                date
            }
        }
    }
`);

export const DELETE_REVIEW = gql(`
    mutation DeleteReview($id: ID!) {
        deleteReview(_id: $id) {
            _id
            vote_average
            vote_count
            reviews {
                _id
                username
                rating
                comment
                date
            }
        }
    }
`);
