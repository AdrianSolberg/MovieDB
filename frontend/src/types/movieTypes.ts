/* eslint-disable @typescript-eslint/no-empty-object-type */

// We disable the eslint rule above because we want to use empty interfaces to
// differentiate between different types even though they have the same structure

interface IdName {
    id: number;
    name: string;
}

interface Genre extends IdName {}
interface Keyword extends IdName {}
interface Language extends IdName {}
interface ProductionCompany extends IdName {}
interface ProductionCountry extends IdName {}

enum Status {
    Released = "Released",
    InProduction = "In Production",
    PostProduction = "Post Production",
    Planned = "Planned",
    Rumored = "Rumored",
    Canceled = "Canceled",
}

interface Movie {
    id: number; // Unique identifier
    title: string; // Unique in combination with release_date
    vote_average: number;
    vote_count: number;
    status: Status;
    release_date: Date;
    revenue: number;
    runtime: number;
    adult: boolean;
    backdrop_path: string | undefined;
    budget: number;
    homepage: string | undefined;
    imdb_id: string | undefined;
    original_language: string;
    original_title: string;
    overview: string;
    popularity: number;
    poster_path: string | undefined;
    tagline: string | undefined;
    genres: Genre[];
    production_companies: ProductionCompany[];
    production_countries: ProductionCountry[];
    spoken_languages: Language[];
    keywords: Keyword[];
}

export { Status };
export type {
    Genre,
    Keyword,
    Language,
    ProductionCompany,
    ProductionCountry,
    Movie,
};
