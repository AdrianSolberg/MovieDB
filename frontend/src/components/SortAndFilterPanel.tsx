import { Button } from "../shadcn/components/ui/button";
import { Accordion } from "../shadcn/components/ui/accordion";
import {
    Sheet,
    SheetContent,
    SheetDescription,
    SheetFooter,
    SheetHeader,
    SheetTitle,
    SheetTrigger,
} from "../shadcn/components/ui/sheet";
import { SlidersHorizontal } from "lucide-react";
import FilterSection from "./FilterSection";
import SortSection from "./SortSection";
import {
    filtersVar,
    searchVar,
    sortOptionVar,
    totalHitsVar,
} from "@/utils/cache";
import { useQuery, useReactiveVar } from "@apollo/client";
import { Filter, Filters, FiltersInput } from "@/types/__generated__/types";
import { GET_FILTERS } from "@/api/queries";
import { defaultSortOption } from "@/utils/sortOptionUtil";
import Loader from "./Loader";
import { useState } from "react";

const SortAndFilterPanel: React.FC = () => {
    const filters = useReactiveVar(filtersVar);
    const search = useReactiveVar(searchVar);
    const [fetchedFilters, setFetchedFilters] = useState<Filters>({
        Genre: [],
        Rating: [],
        Decade: [],
        Status: [],
        Runtime: [],
    });

    const { loading, error } = useQuery(GET_FILTERS, {
        variables: {
            appliedFilters: filters,
            search: search,
        },
        onCompleted: (data) => {
            let totalHits: number;
            // Find the total number of hits
            if (filtersVar().Status.length == 0) {
                // Since Status is union based, we can sum all its hits when no Status is applied
                totalHits = data.filters.Status.map(
                    (s: Filter) => s.hits
                ).reduce((acc, curr) => acc + curr, 0);
            } else {
                // Otherwise, sum only the hits of the applied Status filters
                totalHits = data.filters.Status.filter((s: Filter) =>
                    filtersVar().Status.includes(s.name)
                )
                    .map((s: Filter) => s.hits)
                    .reduce((acc, curr) => acc + curr, 0);
            }
            totalHitsVar(totalHits);

            setFetchedFilters(data.filters);
        },
    });

    const updateFilters = (category: keyof FiltersInput, filter: string) => {
        let newFilters: string[] = [...(filters[category] || [])];
        if (newFilters.includes(filter)) {
            // If the filter is already applied, remove it
            newFilters = newFilters.filter((e) => e != filter);
        } else {
            // Otherwise, add it
            newFilters.push(filter);
        }
        // Update the changed filter category
        const updatedFilters: FiltersInput = {
            ...filters,
            [category]: newFilters,
        };
        sessionStorage.setItem("filters", JSON.stringify(updatedFilters));
        filtersVar(updatedFilters);
    };

    const clearAll = () => {
        const emptyFilters: FiltersInput = {
            Genre: [],
            Rating: [],
            Decade: [],
            Status: [],
            Runtime: [],
        };
        sessionStorage.setItem("filters", JSON.stringify(emptyFilters));
        filtersVar(emptyFilters);

        sessionStorage.setItem("sort_option", defaultSortOption);
        sortOptionVar(defaultSortOption);
    };

    const renderFilterSections = () => {
        return Object.entries(fetchedFilters).map(([category, filter_list]) => {
            if (category === "__typename" || filter_list === "Filters") {
                return null;
            }
            const all_filters = filter_list as Filter[];
            const applied_filters = filters[category as keyof FiltersInput] as
                | string[]
                | undefined;
            return (
                <FilterSection
                    key={category}
                    category={category as keyof FiltersInput}
                    all_filters={all_filters}
                    applied_filters={applied_filters ?? []}
                    updateFilters={updateFilters}
                    loading={loading}
                />
            );
        });
    };

    return (
        <Sheet>
            <SheetTrigger asChild>
                <Button variant="outline" size="default" className="space-x-2">
                    <SlidersHorizontal className="h-3.5 w-3.5" />
                    <span className="sr-only sm:not-sr-only sm:whitespace-nowrap">
                        Sort & Filter
                    </span>
                </Button>
            </SheetTrigger>
            <SheetContent side="left" className="w-full overflow-y-auto">
                <SheetHeader className="mb-5">
                    <SheetTitle>Sort & Filter</SheetTitle>
                    <SheetDescription>
                        Adding filters or changing sorting order will
                        automatically update the results.
                    </SheetDescription>
                </SheetHeader>

                {error ? (
                    <section className="text-center">
                        <p>Something went wrong!</p>
                        <p className="text-primary">Try to refresh</p>
                    </section>
                ) : (
                    <section>
                        <Accordion type="single" collapsible className="w-full">
                            <SortSection loading={loading} />
                            {renderFilterSections()}
                        </Accordion>
                        <SheetFooter className="mt-5">
                            <Button type="reset" onClick={clearAll}>
                                Clear All
                            </Button>
                        </SheetFooter>
                    </section>
                )}
                {loading && (
                    <Loader size="sm">
                        <p>Updating filters...</p>
                    </Loader>
                )}
            </SheetContent>
        </Sheet>
    );
};

export default SortAndFilterPanel;
