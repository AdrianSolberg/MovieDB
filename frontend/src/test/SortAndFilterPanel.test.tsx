import { render, screen } from "@testing-library/react";
import { MockedProvider } from "@apollo/client/testing";
import SortAndFilterPanel from "@/components/SortAndFilterPanel";
import { GET_FILTERS } from "@/api/queries";
import {
    filtersVar,
    searchVar,
    sortOptionVar,
    totalHitsVar,
} from "@/utils/cache";
import { useReactiveVar } from "@apollo/client";
import userEvent from "@testing-library/user-event";
import { vi } from "vitest";
import { SortingType } from "@/types/__generated__/types";
import "@testing-library/jest-dom";

vi.mock("@/utils/cache", () => ({
    filtersVar: vi.fn(),
    searchVar: vi.fn(),
    sortOptionVar: vi.fn(),
    totalHitsVar: vi.fn(),
}));

vi.mock("@apollo/client", async () => {
    const original = await vi.importActual("@apollo/client");
    return {
        ...original,
        useReactiveVar: vi.fn(),
    };
});

vi.mock("@/utils/formatUtil", () => ({
    formatNumber: vi.fn((num: number) => num.toString()),
}));

vi.mock("../components/SortSection", () => ({
    default: vi.fn(({ loading }: { loading: boolean }) => (
        <div data-testid="sort-section">
            <span>Loading: {loading.toString()}</span>
        </div>
    )),
}));

vi.mock("../components/Loader", () => ({
    default: vi.fn(({ children }) => (
        <div data-testid="loader">{children}</div>
    )),
}));

const mockGetFiltersQuery = [
    {
        request: {
            query: GET_FILTERS,
            variables: {
                appliedFilters: {
                    Genre: [],
                    Rating: [],
                    Decade: [],
                    Status: [],
                    Runtime: [],
                },
                search: "",
            },
        },
        result: {
            data: {
                filters: {
                    Genre: [{ name: "Action", hits: 5 }],
                    Rating: [{ name: "4", hits: 10 }],
                    Decade: [{ name: "2010s", hits: 12 }],
                    Status: [{ name: "Released", hits: 15 }],
                    Runtime: [{ name: "1 - 2 hours", hits: 6 }],
                },
            },
        },
    },
    {
        request: {
            query: GET_FILTERS,
            variables: {
                appliedFilters: {
                    Genre: ["Action"],
                    Rating: ["5"],
                    Decade: ["2010s"],
                    Status: ["Released"],
                    Runtime: ["1 - 2 hours"],
                },
                search: "",
            },
        },
        result: {
            data: {
                filters: {
                    Genre: [{ name: "Action", hits: 5 }],
                    Rating: [{ name: "4", hits: 10 }],
                    Decade: [{ name: "2010s", hits: 12 }],
                    Status: [{ name: "Released", hits: 15 }],
                    Runtime: [{ name: "1 - 2 hours", hits: 6 }],
                },
            },
        },
    },
];

const errorMock = [
    {
        request: {
            query: GET_FILTERS,
            variables: {
                appliedFilters: {
                    Genre: [],
                    Rating: [],
                    Decade: [],
                    Status: [],
                    Runtime: [],
                },
                search: "",
            },
        },
        error: new Error("Query error"),
    },
];

describe("SortAndFilterPanel", () => {
    const renderSortAndFilterPanel = async (mock: any) => {
        render(
            <MockedProvider mocks={mock} addTypename={false}>
                <SortAndFilterPanel />
            </MockedProvider>
        );

        await userEvent.click(
            screen.getByRole("button", { name: /Sort & Filter/i })
        );
    };

    beforeEach(() => {
        vi.mocked(filtersVar).mockReturnValue({
            Genre: [],
            Rating: [],
            Decade: [],
            Status: [],
            Runtime: [],
        });
        vi.mocked(searchVar).mockReturnValue("");
        vi.mocked(sortOptionVar).mockReturnValue(SortingType.MOST_POPULAR);
        vi.mocked(totalHitsVar).mockReturnValue(50);
        vi.mocked(useReactiveVar).mockImplementation((varFn) => varFn());
    });

    afterEach(() => {
        vi.clearAllMocks();
        sessionStorage.clear();
    });

    it("renders Sort & Filter button", () => {
        render(
            <MockedProvider mocks={mockGetFiltersQuery} addTypename={false}>
                <SortAndFilterPanel />
            </MockedProvider>
        );

        expect(
            screen.getByRole("button", { name: /Sort & Filter/i })
        ).toBeInTheDocument();
    });

    it("displays filter sections and Clear All button", async () => {
        await renderSortAndFilterPanel(mockGetFiltersQuery);

        expect(screen.getByText(/Genre/i)).toBeInTheDocument();
        expect(screen.getByText(/Rating/i)).toBeInTheDocument();
        expect(screen.getByText(/Decade/i)).toBeInTheDocument();
        expect(screen.getByText(/Status/i)).toBeInTheDocument();
        expect(screen.getByText(/Runtime/i)).toBeInTheDocument();

        expect(screen.getByText(/Clear All/i)).toBeInTheDocument();
    });

    it("adds filter when an option is clicked", async () => {
        await renderSortAndFilterPanel(mockGetFiltersQuery);

        await userEvent.click(screen.getByRole("button", { name: /Genre/i }));

        await userEvent.click(
            screen.getByRole("checkbox", { name: /Action/i })
        );
        expect(filtersVar).toHaveBeenCalledWith({
            Genre: ["Action"],
            Rating: [],
            Decade: [],
            Status: [],
            Runtime: [],
        });
        expect(sessionStorage.getItem("filters")).toBe(
            JSON.stringify({
                Genre: ["Action"],
                Rating: [],
                Decade: [],
                Status: [],
                Runtime: [],
            })
        );
    });

    it("removes filter when an option is clicked", async () => {
        vi.mocked(filtersVar).mockReturnValue({
            Genre: ["Action"],
            Rating: ["5"],
            Decade: ["2010s"],
            Status: ["Released"],
            Runtime: ["1 - 2 hours"],
        });

        await renderSortAndFilterPanel(mockGetFiltersQuery);

        await userEvent.click(screen.getByRole("button", { name: /Genre/i }));

        await userEvent.click(
            screen.getByRole("checkbox", { name: /Action/i })
        );

        expect(filtersVar).toHaveBeenCalledWith({
            Genre: [],
            Rating: ["5"],
            Decade: ["2010s"],
            Status: ["Released"],
            Runtime: ["1 - 2 hours"],
        });
        expect(sessionStorage.getItem("filters")).toBe(
            JSON.stringify({
                Genre: [],
                Rating: ["5"],
                Decade: ["2010s"],
                Status: ["Released"],
                Runtime: ["1 - 2 hours"],
            })
        );
    });

    it("clears all filters when Clear All button is clicked", async () => {
        vi.mocked(filtersVar).mockReturnValue({
            Genre: ["Action"],
            Rating: ["5"],
            Decade: ["2010s"],
            Status: ["Released"],
            Runtime: ["1 - 2 hours"],
        });

        await renderSortAndFilterPanel(mockGetFiltersQuery);

        await userEvent.click(screen.getByText(/Clear All/i));

        expect(filtersVar).toHaveBeenCalledWith({
            Genre: [],
            Rating: [],
            Decade: [],
            Status: [],
            Runtime: [],
        });
        expect(sortOptionVar).toHaveBeenCalledWith(SortingType.MOST_POPULAR);
        expect(sessionStorage.getItem("filters")).toBe(
            JSON.stringify({
                Genre: [],
                Rating: [],
                Decade: [],
                Status: [],
                Runtime: [],
            })
        );
        expect(sessionStorage.getItem("sort_option")).toBe(
            SortingType.MOST_POPULAR
        );
    });

    it("displays an error message if query fails", async () => {
        await renderSortAndFilterPanel(errorMock);

        expect(screen.getByText(/Something went wrong!/i)).toBeInTheDocument();
        expect(screen.getByText(/Try to refresh/i)).toBeInTheDocument();
    });
});
