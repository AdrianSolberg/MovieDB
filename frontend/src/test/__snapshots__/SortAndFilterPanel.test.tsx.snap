// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SortAndFilterPanel > matches snapshot 1`] = `
<DocumentFragment>
  <button
    aria-controls="radix-:r0:"
    aria-expanded="false"
    aria-haspopup="dialog"
    class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2 space-x-2"
    data-state="closed"
    type="button"
  >
    <svg
      class="lucide lucide-sliders-horizontal h-3.5 w-3.5"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <line
        x1="21"
        x2="14"
        y1="4"
        y2="4"
      />
      <line
        x1="10"
        x2="3"
        y1="4"
        y2="4"
      />
      <line
        x1="21"
        x2="12"
        y1="12"
        y2="12"
      />
      <line
        x1="8"
        x2="3"
        y1="12"
        y2="12"
      />
      <line
        x1="21"
        x2="16"
        y1="20"
        y2="20"
      />
      <line
        x1="12"
        x2="3"
        y1="20"
        y2="20"
      />
      <line
        x1="14"
        x2="14"
        y1="2"
        y2="6"
      />
      <line
        x1="8"
        x2="8"
        y1="10"
        y2="14"
      />
      <line
        x1="16"
        x2="16"
        y1="18"
        y2="22"
      />
    </svg>
    <span
      class="sr-only sm:not-sr-only sm:whitespace-nowrap"
    >
      Sort & Filter
    </span>
  </button>
</DocumentFragment>
`;
